swarm/storage/mru/lookup/lookup_test.go

```diff
@@ -0,0 +1,324 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package lookup_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/swarm/storage/mru/lookup"
+)
+
+type Data struct {
+	Payload uint64
+	Time    uint64
+}
+
+type Store map[lookup.EpochID]*Data
+
+func write(store Store, epoch lookup.Epoch, value *Data) {
+	fmt.Printf("Write: %d-%d, value='%d'\n", epoch.Base(), epoch.Level, value.Payload)
+	store[epoch.ID()] = value
+}
+
+func update(store Store, last lookup.Epoch, now uint64, value *Data) lookup.Epoch {
+	var epoch lookup.Epoch
+
+	epoch = lookup.GetNextEpoch(last, now)
+
+	write(store, epoch, value)
+
+	return epoch
+}
+
+const Day = 60 * 60 * 24
+const Year = Day * 365
+const Month = Day * 30
+
+func makeReadFunc(store Store, counter *int) lookup.ReadFunc {
+	return func(epoch lookup.Epoch, now uint64) (interface{}, error) {
+		*counter++
+		data := store[epoch.ID()]
+		var valueStr string
+		if data != nil {
+			valueStr = fmt.Sprintf("%d", data.Payload)
+		}
+		fmt.Printf("Read: %d-%d, value='%s'\n", epoch.Base(), epoch.Level, valueStr)
+		if data != nil && data.Time <= now {
+			return data, nil
+		}
+		return nil, nil
+	}
+}
+
+func TestLookup(t *testing.T) {
+
+	store := make(Store)
+	readCount := 0
+	readFunc := makeReadFunc(store, &readCount)
+
+	// write an update every month for 12 months 3 years ago and then silence for two years
+
+	now := uint64(1533799046)
+	var epoch lookup.Epoch
+
+	var lastData *Data
+	for i := uint64(0); i < 12; i++ {
+		t := uint64(now - Year*3 + i*Month) // update every month for 12 months 3 years ago and then silence for two years
+		data := Data{
+			Payload: t, //our "payload" will be the timestamp itself.
+			Time:    t,
+		}
+		epoch = update(store, epoch, t, &data)
+		lastData = &data
+	}
+
+	// try to get the last value
+
+	value, err := lookup.Lookup(now, lookup.NoClue, readFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	readCountWithoutHint := readCount
+
+	if value != lastData {
```

Comment:
why compare pointers here?

Reply:
It is sufficient to compare pointers in this test because of how the test is laid out, only one instance for each item.